File: app/src/main/java/io/github/getsmp/lorforandroid/ui/section/forum/section/ForumSectionItemFactory.java
/*
 * Copyright 2016 getsmp
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.getsmp.lorforandroid.ui.section.forum.section;

import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.util.List;

import io.github.getsmp.lorforandroid.ui.section.ItemCommon;
import io.github.getsmp.lorforandroid.ui.section.ItemFactory;
import io.github.getsmp.lorforandroid.util.StringUtils;

public class ForumSectionItemFactory implements ItemFactory {
    @Override
    public void prepareItems(Element body, List items) {
        Elements entries = body.select("tbody tr");
        for (Element entry : entries) {
            Element properties = entry.select("td").first();
            String bareAuthor = properties.ownText();
            items.add(new ItemCommon(
                    properties.select("a").first().attr("href").substring(1),
                    properties.select("a").first().ownText(),
                    null,
                    StringUtils.tagsFromElements(properties.select("a").first().select("span.tag")),
                    entry.select("td.dateinterval").first().select("time").first().ownText(),
                    bareAuthor.substring(bareAuthor.lastIndexOf("("), bareAuthor.lastIndexOf(")")).replaceAll("[()]", ""),
                    StringUtils.addEnding(entry.select("td.numbers").first().ownText())
            ));
        }
    }
}

File: app/src/main/java/io/github/getsmp/lorforandroid/ui/section/forum/section/ForumSectionItemFactory.kt
/*
 * Copyright (c) 2016 getsmp
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package io.github.getsmp.lorforandroid.ui.section.forum.section

import io.github.getsmp.lorforandroid.ui.section.ItemFactory
import io.github.getsmp.lorforandroid.util.StringUtils
import org.jsoup.nodes.Element

class ForumSectionItemFactory : ItemFactory {
    override fun prepareItems(body: Element, items: MutableList<Any>) {
        val entries = body.select("tbody tr")
        for (entry in entries) {
            val properties = entry.select("td").first()
            val bareAuthor = properties.ownText()
            items.add(ForumSectionItem(
                    url = properties.select("a").first().attr("href").substring(1),
                    title = properties.select("a").first().ownText(),
                    groupTitle = null,
                    tags = StringUtils.tagsFromElements(properties.select("a").first().select("span.tag")),
                    date = entry.select("td.dateinterval").first().select("time").first().ownText(),
                    author = bareAuthor.substring(bareAuthor.lastIndexOf("("), bareAuthor.lastIndexOf(")")).replace("[()]".toRegex(), ""),
                    comments = StringUtils.numericStringToHumanReadable(entry.select("td.numbers").first().ownText()),
                    isPinned = properties.select("i.icon-pin").size > 0
            ))
        }
    }
}
